ThePhasesOfCraftship/2_best_practice_first/testingBasics/exercises/2_Arrange_Act_Assert_Backwards/2_1_Password_Validator/src/index.spec.ts

```diff
@@ -1,9 +1,120 @@
+import { PasswordValidator, ValidationResult, ValidationError, ValidationErrorType } from ".";
+
+// Utility function for asserting successful validation
+function expectToPass(validationResult: ValidationResult) {
+  expect(validationResult.result).toBe(true);
+  expect(validationResult.errors).toHaveLength(0);
+}
+
+// Utility function for asserting failed validation
+function expectToFail(validationResult: ValidationResult, ...errors: ValidationError[]) {
+  expect(validationResult.result).toBe(false);
+  expect(validationResult.errors).toHaveLength(errors.length);
+  expect(validationResult.errors).toMatchObject(errors);
+}
```

Comment:
Hello Abel. I see you are striving for perfection in code but this really is an overkill and I think we can remove this because we are basically introducing abstraction over jest which IMO is not necessary

Reply:
Hi!
I respectfully disagree with this part. Keeping your test code clean, readable, duplication-free can help a great deal on the long run. Uncle Bob in Clean Code Chapter 9: Unit tests - Clean Tests uses similar approaches to make the test code cleaner. I'm not saying that we should blindly accept what the old-school masters say, but give it a try there's usually some truth in it 🙂

Through my journey as a FE dev I found that test utility functions like these can help you to avoid the ripple effect that could happen if your API changes. Ex. if the structure of the `ValidationResult` type changes drastically, I'd have to make the same changes in my test code all over the place. With this approach, I only have to change it at 2 places.

I found them exceptionally helpful when writing unit tests for FE components. Ex. if you were to test the interactions with a select component you might use something like this to verify the currently selected option:

```ts
it('displays the currently selected item', () => {
    const handleChange = jest.fn();
    
    // Render the component with an initially selected value
    render(
      <SingleSelect
        options={options}
        value="2"
        onChange={handleChange}
        label="Select an option"
        placeholder="Choose one"
      />
    );

    // Open the dropdown
    fireEvent.click(screen.getByText(/Choose one/i));

    // Check that the correct item is displayed as selected
    const selectedOption = screen.getByText('Option 2');
    expect(selectedOption).toBeInTheDocument();

    // Check that the selected item has the 'selected' class
    expect(selectedOption).toHaveClass('selected');

    // Ensure that the dropdown is visible
    const dropdown = screen.getByRole('listbox');
    expect(dropdown).toBeVisible();
  });
});
```

Compare it with the following version. I think there's no doubt which one is cleaner.

```ts
it('displays the currently selected item', () => {
    const handleChange = jest.fn();
    
    // Render the component with an initially selected value
    render(
      <SingleSelect
        options={options}
        value="2"
        onChange={handleChange}
        label="Select an option"
        placeholder="Choose one"
      />
    );

    openDropdown();

    verifySelectedOption('Option 2');

    verifyDropdownOpen();
  });
});
```

And probably you'll have more tests like these if you want to cover all interactions with the component.